### DISCUSSION TITLE
How to include local dependency

### DISCUSSION THREAD

**Discussion by simse07** [dasniko/testcontainers-keycloak#57] [General]
Hi, thank you for the great project.

How can I include my local dependency together with the extension I want to test?

Example:
- extension
   - src/main/...
   - src/test/...
- another-extension
   - src/main/... 
- core
   - src/main/... 

I want to test my `extension`, which has a dependency to my `core` package e.g. some exception or util classes.
When I start the testcontainer test in the `extension/src/test` while passing `withExtensionClassesFrom("target/classes")` I get an error `org.jboss.modules.ModuleNotFoundException: core.exception`.

Maybe we can define multiple paths for `withExtensionClassesFrom`?

**Comment by dasniko** [comment]
Hi @simse07, thanks for coming up with this, sounds like a good idea.

I don't know if it's a good idea to include several different `classes` folders, but I will think about.
It may work in your usecase, but others might some classpath errors in some cases!?

What I could think of something like `withExtensionLibsFrom(...)`, this would empower to include 3rd party libs together with an extension.

WDYT?

**Comment by simse07** [comment]
Thanks for the fast reply.
`withExtensionLibsFrom(...)` sounds like a good extension to me.

**Comment by simse07** [comment]
I was thinking about this and probably I have expressed myself in a misleading way.

The extension lib should be added as module to the Keycloak. So I would rather name it like `addCustomModule(...)` where you have to specify the name & package of the lib, path to the jar and to the module.xml:

`addCustomModule("core", "my.package.core", "../core/target/core.jar", "../core/module.xml")`

This will be deployed to `/opt/jboss/keycloak/modules/my/package/core/main/core.jar` and `/opt/jboss/modules/my/package/core/main/module.xml`.

What do you think?

**Comment by dasniko** [comment]
Ah, seems that you reference to the v1.x version of this project and thus to Wildfly-based Keycloak.
As the Wildfly distribution is now deprecated and the Quarkus-based distribution is the default, I also deprecated the v1.x version of this project. V2.x uses the Quarkus-based Keycloak. And there, you don't have modules.
Unfortunately (? no, not really), I won't add new features to the v1.x branch, only 2.x is now supported.

However, I will add some feature like I mentioned for libs to one of the next releases.

**Comment by simse07** [comment]
I was hoping that you would not say that, but totally understandable.

For others who would like to add a module to v1.x:
```
public class ModuleDto {
    String moduleName; // core
    String modulePackage; // my.package.core, same as the name in the module.xml
    String moduleJarPath; // ../core/target/core.jar
    String modulePath; // ../core/module.xml
}

/* ----------------------------- */

public class KeycloakContainerModule extends KeycloakContainer {
    private ModuleDto moduleDto;
   
    public KeycloakContainerModule withModuleDeployment(ModuleDto moduleDto) {
        this.moduleDto = moduleDto;
        return (KeycloakContainerModule) self();
    }

    @Override
    protected void configure() {
        super.configure();
        if (moduleDto != null) {
            createKeycloakModuleDeployment(moduleDto);
        }
    }


    public void createKeycloakModuleDeployment(ModuleDto moduleDto) {
        String deploymentLocation = "/opt/jboss/keycloak/modules/" + moduleDto.modulePackage.replace(".", "/") + "/main/";
        withCopyFileToContainer(MountableFile.forHostPath(Paths.get(moduleDto.moduleJarPath)), deploymentLocation + moduleDto.moduleName + ".jar");
        withCopyFileToContainer(MountableFile.forHostPath(Paths.get(moduleDto.modulePath)), deploymentLocation + "module.xml");
    }
}
```

**Comment by dasniko** [comment]
Nice, thanks. 🙂
If you don't mind, I will convert this issue to a discussion thread, so it is saved for others.
I'll create another issue to add the `withExtensionLibsFrom()` as discussed above.

**Reply by dasniko** [reply]
Hey @simse07 
I just pushed the new feature. There's now the method `.withProviderLibsFrom(List<File> files)` which adds the given list of files as dependencies. See also the README and tests for usage with Maven dependencies.
But, as already mentioned, it's only in the 2.x version.

It's not yet published, but you can use a snapshot dependency via https://jitpack.io/#dasniko/testcontainers-keycloak/main-SNAPSHOT

**Reply by simse07** [reply]
Great, thank you very much. I will test it as soon as we upgraded to Keycloak 17.
